File: perl/tests/TestsTrie.pl
#!/usr/bin/env perl
use FiftyOneDegrees::TrieV3;
use feature qw/say/;
use Time::HiRes qw/ time /;
use Memory::Usage;

my $litefilename = "../../data/51Degrees-LiteV3.2.trie";
my $premiumfilename = "../../data/51Degrees-PremiumV3.2.trie";
my $enterprisefilename = "../../data/51Degrees-EnterpriseV3.2.trie";
my $inputFile = "../../data/20000 User Agents.csv";
my $propertyList = "IsMobile";
my $cacheSize = 10000;
my $poolSize = 20;

sub speedtest{
	my $provider = @_[0];
	my $file_in = @_[1];
	open my $file_in, $inputFile;
	my $speedtest_start = time;
	while( my $userAgent = <$file_in>) {
		chomp $userAgent;
		my $match = $provider->getMatch($userAgent);
		last if $. == 20000;
	}
	my $speedtest_end = time;
	return sprintf("%.3f", ( $speedtest_end - $speedtest_start ) / 20);
	close $file_in;
}

sub runtests{
	my $dataFile = @_[0];
	if(-e $dataFile){
		my $mu = Memory::Usage->new();
		$mu->record('start');
		my $start = time;

		my $provider = new FiftyOneDegrees::TrieV3::Provider(
			$dataFile,
			$propertyList);
		my $end = time;

		$mu->record('end');

		my $startmem = @{@{ $mu->state }[0]}[3];
		my $endmem =  @{@{ $mu->state }[1]}[3];
		my $memused = ($endmem - $startmem) / 1000;

		print "   Initialization time: " . ( $end - $start)*1000 . " ms\n";
		print "   Detections speed: " . speedtest($provider, $inputfile) . " ms\n";
		print "   Memory usage: " . $memused . " Mb\n";
	}
	else{
		print "   ".$dataFile." could not be found.\n";
	}
}

print "Lite Tests:\n";
runtests($litefilename);

print "Premium Tests:\n";
runtests($premiumfilename);

print "Enterprise Tests:\n";
runtests($enterprisefilename);

File: python/tests/TestsTrie.py
#!/usr/bin/env python

from FiftyOneDegrees import fiftyone_degrees_mobile_detector_v3_trie_wrapper
from fiftyone_degrees.mobile_detector.conf import settings
import sys
import os
import time
import resource

liteDataFile = "../../data/51Degrees-LiteV3.4.trie"
premiumDataFile = "../../data/51Degrees-PremiumV3.4.trie"
enterpriseDataFile = "../../data/51Degrees-EnterpriseV3.4.trie"
inputFile = '../../data/20000 User Agents.csv'
properties = settings.PROPERTIES

def speedtest(provider):
	fin = open(inputFile, 'r')
	speedtest_start = time.time()
	i = 0
	while i < 20000 :
		userAgent = fin.readline().rstrip('\n')
		match = provider.getMatch(userAgent)
		i += 1
	speedtest_end = time.time()
	fin.close()
	return (speedtest_end - speedtest_start)/(i/1000)

def runtests(dataFile):
	if (os.path.isfile(dataFile)):
		start = time.time()
		provider = fiftyone_degrees_mobile_detector_v3_trie_wrapper.Provider(
			dataFile,
			properties)

		end = time.time()

		print "   Initialization time: %s ms" % ((end - start)*1000)
		print "   Time per detection: %s ms" % speedtest(provider)
		print "   Memory usage: %d Mb" % (resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/1000)
		del(provider)
	else:
		print "   %s could not be found." % dataFile

print "Lite Tests:"
runtests(liteDataFile)

print "Premium Tests:"
runtests(liteDataFile)

print "Enterprise Tests:"
runtests(enterpriseDataFile)
